scripts/building/check_enforcer.ps1

```diff
@@ -129,29 +121,78 @@ function sdk_cli_checks([ref]$failed_reason_ref, $valid_status_array) {
     }
 }
 
-function run_checks() {
-
-    $need_to_check = New-Object System.Collections.Generic.HashSet[string]
-    git diff --name-only HEAD origin/main | ForEach-Object {
-        if ($snippet_debug -eq 1) {
-            Write-Output "git diff --name-only HEAD main $_"
+function trigger_prepare($input_paths, [ref]$failed_reason_ref) {
+    foreach ($input_path in $input_paths) {
+        if ($checks.ContainsKey("samples_connections_connection")) {
+            continue
+        }
+        if ($input_path.Contains("examples") -or $input_path.Contains("samples")) {
+            # Define the input path
+            Push-Location $GithubWorkspace
+            $pipelines_samples = (python $GithubWorkspace/scripts/readme/readme.py -c | ConvertFrom-Json)
+            Pop-Location
+            git diff --name-only HEAD | ForEach-Object {
+                $failed_reason_ref.Value = "Run readme generation before check in"
+                return
+            }
+            if ($failed_reason_ref.Value -ne "") {
+                return
+            }
+            # merge piplines_samples to checks
+            foreach ($key in $pipelines_samples.psobject.properties.name) {
+                $value = $pipelines_samples.$key
+                $checks[$key] = $value
+            }
         }
-        if ($_.Contains("src/promptflow/")) {
-            $need_to_check.Add("sdk_cli")
+    }
+    # reverse checks
+    foreach ($key in $checks.Keys) {
+        $value = $checks[$key]
+        foreach ($path in $value) {
+            if ($reverse_checks.ContainsKey($path)) {
+                $reverse_checks[$path] += $key
+            } else {
+                $reverse_checks[$path] = @($key)
+            }
         }
     }
-
-    $sdk_cli_check = $false
-
-    foreach ($item in $need_to_check) {
-        if ($item -eq "sdk_cli") {
-            $sdk_cli_check = $true
+    # render $pipelines and $pipelines_count using $input_paths
+    foreach ($input_path in $input_paths) {
+        # input pattern /**: input_path should match in the middle
+        # input pattern /*: input_path should match last but one
+        # other input pattern: input_path should match last
+        $keys = $reverse_checks.Keys
+        $keys = $keys | Where-Object { (python $GithubWorkspace/scripts/building/fnmatch.py -g $_ -f $input_path) -eq "True" }
+        # $reverse_checks.Keys and $keys
+        foreach ($key_item in $keys) {
+            foreach ($key in $reverse_checks[$key_item]) {
+                if ($pipelines.ContainsKey($key)) {
+                    if ($special_care.ContainsKey($key)) {
+                        $pipelines[$key] = $special_care[$key]
```

Comment:
intuitively, we need a preprocess, to make the key as path, and value is the triggered pipeline, so that this loop can be much smaller?

Reply:
Changed.